### PR TITLE
An implementation of checkpoints

### LINKED ISSUE #17: Add checkpoints to build_from.py
Right now, building a regulation is incredibly painful, particularly when something goes wrong. This is in large part because re-running the _working_ steps takes so long.

Add the ability for build_from.py to checkpoint its progress so that future runs don't require as much rework. As part of this, merge in https://github.com/cfpb/regulations-parser/pull/262

### PR DESCRIPTION
**Pull request by cmc333333**
This will hopefully resolve #17 by pickling expensive-to-create data structures throughout the build_from pipeline. Specifically, a checkpoint is saved for

1. the initial xml tree parsing
1. determination of the tree's notice number
1. parsing all of the notices (as a group -- probably want to split this up later)
1. building each of the layers
1. compiling each version of the regulation
1. each generated diff

Also merged in @grapesmoker's very nice argument parsing improvements and updated the README.

So far, I've only tested this on FEC's 110 and CFPB's reg E, but I think it should work fine for CFPB's reg Z, too.

### CODE REVIEW

**Comment by cmc333333** [comment]
To show how this is useful:

I'm trying to debug a reg Z error I've been getting after it starts to compile the 10th version of the reg. So, I run this once:
```bash
python build_from.py fr-notices/annual/CFR-2012-title12-vol8-part1026.xml 12 15 1693 --checkpoint cps
``` 
which dies after about a half hour. If I run the command again, it'll die at the same point after around thirty seconds. Combine that with ipdb:

```bash
ipython --pdb --c "%run build_from.py fr-notices/annual/CFR-2012-title12-vol8-part1026.xml 12 15 1693 --checkpoint cps"
```
and now I'm dropped into a debugger after thirty seconds, to try to figure out what's wrong with the notice:

```python
AttributeError: 'NoneType' object has no attribute 'text'
> /home/vagrant/regulations-parser/regparser/notice/compiler.py(366)replace_node_text()
    365         node = find(self.tree, label)
--> 366         node.text = change['node']['text']
    367

ipdb>
```

**Comment by cmc333333** [comment]
@ascott1 @grapesmoker @willbarton - would also love your input here. It'd make the PR upstream to CFPB much easier to not need to tweak it for feedback twice :)

**Comment by khandelwal** [comment]
Thoughts: Pickle is a binary format right? Does it make more sense for these checkpoints to be in a text-based format like JSON? That allows inspection through text editor, grep or other custom programs. That might allow us to build better tooling around the Regulation Build System. I feel like that would be more useful in the long run. 

On the other hand, if one can load a pickle file into a debugger and with relatively little effort manipulate and inspect the contents, that might help.

**Comment by grapesmoker** [comment]
Sorry, haven't had time to look at this at all. Is there any reason we
can't have both JSON and pickle, one for human readability and one for easy
loading into debugger?

On Mon, Apr 20, 2015 at 10:31 AM, Shashank Khandelwal <
notifications@github.com> wrote:

> Thoughts: Pickle is a binary format right? Does it make more sense for
> these checkpoints to be in a text-based format like JSON? That allows
> inspection through text editor, grep or other custom programs. That might
> allow us to build better tooling around the Regulation Build System. I feel
> like that would be more useful in the long run.
>
> On the other hand, if one can load a pickle file into a debugger and with
> relatively little effort manipulate and inspect the contents, that might
> help.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/18F/regulations-parser/pull/18#issuecomment-94468025>.
>



-- 
http://www.google.com/profiles/grapesmoker

**Comment by cmc333333** [comment]
The problem with JSON is that it is not aware of Python classes. That means we'd need a converter for every type of thing we're checkpointing, and as that list isn't explicitly defined yet, it seems like a better idea to hold off for a bit. Once we know, switching to JSON should be easy.

**Comment by willbarton** [comment]
Since this is internal state, I think pickling makes the most amount of sense right now, particularly with respect to debugability. Human readability might be nice, but I'm skeptical of it being more useful than a debugger.

Successfully parsed Reg J, second run uses the checkpoints, removing a given checkpoint recreates it, and the output is consistent. Reg Z fails and it's quite nice to get right back to the point where it failed in a debugger. :+1:  

In thinking about this beyond just debugging... if I'm working on a particular part of the pipeline, it might be useful to be able to specify a particular stage before which checkpoints should be used. It might also be worth being able to determine checkpoint staleness. Both of these things can be accomplished easily enough with `rm`, but these are "might be nice" things that occur to me.

File: regparser/builder.py [hunk 1]
```diff
@@ -216,3 +226,81 @@ class EmptyCache(object):
...
+
+class Checkpointer(object):
+    """Save checkpoints during the build pipeline. Generally, a caller will
+    specify, a unique tag (a string) and a fallback function (for how to
+    compute it when there is no checkpoint). Calling checkpoint increment the
+    counter field, which is prefixed to the filename to limit the risk of
+    re-ordering collisions."""
+    def __init__(self, path):
+        self.counter = 0
+        self.path = path
```
Comment (khandelwal) [inline regparser/builder.py]: Naming this better might help indicate what this is the path to.

File: regparser/builder.py [hunk 2]
```diff
@@ -216,3 +226,81 @@ class EmptyCache(object):
...
+        self.suffix = ""
+        self.ignore_checkpoints = False
+        if not os.path.isdir(path):
+            os.makedirs(path)
+
+    def _filename(self, tag):
```
Comment (khandelwal) [inline regparser/builder.py]: I see some higher level tests below, I'd encourage you to write more tests around some of these lower-level functions.

**Comment by cmc333333** [comment]
Updated to incorporate feedback

**Comment by khandelwal** [comment]
There's a merge conflict here. Once that's resolved, I'll merge this.

**Comment by cmc333333** [comment]
Merged master

**Pull request merged**